### PR TITLE
[fbrp] conda shared env

### PR DESCRIPTION
**Pull request by lshamis**
# Description

Adds up flag `--cache/--nocache`, with cache enabled by default.
* For Docker, `nocache` will force a pull the named image. `nocache` will also be passed to the build command.
* For Conda, `cache` checks if `~/.config/fbrp/conda/{name}.yaml` exists and has the same contents as desired. If so, checks that the conda env history is just the build command. Does a full rebuild otherwise.
  * Takes about ~0.5 sec to check if cache is valid.

Adds `fbrp.Conda.SharedEnv` for shared environments across processes.
```py
shared_env = fbrp.Conda.SharedEnv(
    "shared_env",
    dependencies=["python>=3.7"],
)

for i in range(10):
    fbrp.process(
        name=f"proc_{i}",
        runtime=fbrp.Conda(
            shared_env=shared_env,
            run_command=["python3", "proc.py", i],
        ),
    )
```
(0.5 sec cache check for the shared_env, rather than 0.5 sec per-process)

Rename runtime conda arg `env`->`copy_named_env` and `env_nosandbox`->`use_named_env`

Switches print statements to use `click.echo`. See https://click.palletsprojects.com/en/8.0.x/utils/

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [x] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Before:
Restarting a conda process may take more than a minute, even if the environment hasn't changed.

After:
Restarting a conda process takes less than a second.

# Testing

Manual testing.

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

File: fbrp/src/fbrp/__init__.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 from importlib.machinery import SourceFileLoader
 import click
 import os
+import sys
```
Comment (1heart) [inline fbrp/src/fbrp/__init__.py:9]: Extraneous except for the commented code, I think?

```suggestion
```

File: fbrp/src/fbrp/runtime/conda.py [hunk 2]
```diff
@@ -174,107 +172,187 @@ async def handle_down(self):
...
+                    conda activate {self.use_named_env}
+                """,
+                shell=True,
+                executable="/bin/bash",
+                stderr=subprocess.PIPE,
+            )
+            if result.returncode:
+                raise RuntimeError(f"'use_named_env' not valid: {result.stderr}")
+
+        def _generate_env_content(self, root: pathlib.Path):
```
Comment (1heart) [inline fbrp/src/fbrp/runtime/conda.py:217]: It's not entirely clear at first glance what this method does -- mind adding a docstring?

File: fbrp/src/fbrp/runtime/conda.py [hunk 3]
```diff
@@ -174,107 +172,187 @@ async def handle_down(self):
...
+                    )
+                )
+                history_file = os.path.join(info["prefix"], "conda-meta/history")
+                history = open(history_file).readlines()
+                cmds = [line.strip() for line in history if line.startswith("# cmd: ")]
+                if len(cmds) != 1:
+                    return False
+
+                if not cmds[0].endswith(f"update --prune -f {yaml_path}"):
+                    return False
```
Comment (1heart) [inline fbrp/src/fbrp/runtime/conda.py:262]: Can you document what this does?

File: fbrp/src/fbrp/runtime/conda.py [hunk 4]
```diff
@@ -174,107 +172,187 @@ async def handle_down(self):
...
+        use_named_env=None,
+        copy_named_env=None,
+        shared_env=None,
         yaml=None,
-        env=None,
-        env_nosandbox=None,
         channels=[],
         dependencies=[],
         setup_commands=[],
         use_mamba=None,
```
Comment (1heart) [inline fbrp/src/fbrp/runtime/conda.py:315]: Document these fields or add a typehint

**Review by 1heart — COMMENTED** [review]
Excited for this functionality!

The unit test is a bit basic, and this functionality is pretty important. Maybe it's time to add an integration tests?

File: fbrp/src/fbrp/__init__.py (hunk 1 above)
Comment (lshamis) [reply fbrp/src/fbrp/__init__.py]: Done

File: fbrp/src/fbrp/runtime/conda.py (hunk 3 above)
Comment (lshamis) [reply fbrp/src/fbrp/runtime/conda.py:305]: Added docstring comments. Let me know if it's easier to understand.

File: fbrp/src/fbrp/runtime/conda.py (hunk 2 above)
Comment (lshamis) [reply fbrp/src/fbrp/runtime/conda.py]: Done

File: fbrp/src/fbrp/runtime/conda.py (hunk 4 above)
Comment (lshamis) [reply fbrp/src/fbrp/runtime/conda.py:359]: Added docstring.

**Review by 1heart — APPROVED** [review]
Thanks!

**Pull request merged**